### PR TITLE
feat(ui): add prerequisites in targeting

### PR DESCRIPTION
**Pull request by bimalgrg519**
Solve https://github.com/bucketeer-io/bucketeer/issues/180

### CODE REVIEW

**Comment by cre8ivejp** [comment]
@bimalgrg519 

Every time I change the dropdown menu for the prerequisite or delete an item, it makes a `ListFeatures` request.
It should only request once when the user clicks on the `Add prerequisite`.

- Request `ListFeatures` when there are no prerequisites and the user clicks on the `Add prerequisite`
- Request `ListFeatures` when the page is loaded and there are one or more prerequisites

Also, the feature flag ID and the variation are not being validated.

**Review by cre8ivejp — CHANGES_REQUESTED** [review]

File: ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx [hunk 1]
```diff
@@ -107,6 +113,12 @@ export const FeatureTargetingForm: FC<FeatureTargetingFormProps> = memo(
                 />
               )}
             />
+            <div>
+              <label className="input-section-label">Prerequisites</label>
```
Comment (cre8ivejp) [inline ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:117]: Please add the `Prerequisites` to the lang files.

**Review by cre8ivejp — CHANGES_REQUESTED** [review]

File: ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx (hunk 1 above)
Comment (cre8ivejp) [reply ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:117]: Please use this `前提条件` for the Japanese.

Comment (bimalgrg519) [reply ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:117]: Could you please provide for :

- Add Prerequisite
- Select a flag
- Select a variation

Comment (cre8ivejp) [reply ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:117]: @bimalgrg519 

Please use the following.

- Add Prerequisite (前提条件を追加)
- Select a flag (フィーチャーフラグを選択)
- Select a variation (バリエーションを選択)

**Comment by bimalgrg519** [comment]
> Also, the feature flag ID and the variation are not being validated.

What type of validation should be applied for this?

**Comment by cre8ivejp** [comment]
@bimalgrg519 

> What type of validation should be applied for this?

All the fields that are configured in the request and in the commands (AddPrerequisiteCommand, ChangePrerequisiteVariationCommand, etc) should be validated.

File: ui/web-v2/apps/admin/src/assets/lang/en.json [hunk 2]
```diff
@@ -234,9 +235,12 @@
   "feature.list.noResult.searchKeyword": "ID, name and description",
   "feature.offVariation": "off variation",
   "feature.onVariation": "on variation",
+  "feature.prerequisites": "Prerequisites",
   "feature.resetRandomSampling": "Reset random sampling",
   "feature.rule": "Rule rollout percentage",
   "feature.search.placeholder": "ID, Name, Description",
+  "feature.selectFlag": "Select a flag",
```
Comment (cre8ivejp) [inline ui/web-v2/apps/admin/src/assets/lang/en.json:242]: Please change it to `Select a feature flag`

File: ui/web-v2/apps/admin/src/pages/feature/targeting.tsx [hunk 3]
```diff
@@ -227,13 +253,14 @@ export const FeatureTargetingPage: FC<FeatureTargetingPageProps> = memo(
...
     }, [dispatch, currentEnvironment]);
 
-    if (isLoading) {
-      return (
-        <div className="p-9 bg-gray-100">
-          <DetailSkeleton />
-        </div>
-      );
-    }
+    // if (isLoading) {
```
Comment (cre8ivejp) [inline ui/web-v2/apps/admin/src/pages/feature/targeting.tsx:256]: Is there a reason to comment this code?

**Comment by cre8ivejp** [comment]
@bimalgrg519 

We should disable the `Add prerequisites` button when there are feature flags to add.

Screen Shot 2023-02-08 at 14 03 54

**Comment by cre8ivejp** [comment]
@bimalgrg519 

Can you remove the white background when there are no prerequisites like in the Rule section?

Screen Shot 2023-02-08 at 14 15 13

File: ui/web-v2/apps/admin/src/pages/feature/targeting.tsx (hunk 3 above)
Comment (bimalgrg519) [reply ui/web-v2/apps/admin/src/pages/feature/targeting.tsx:256]: I will remove this. We don't need this anymore.

File: ui/web-v2/apps/admin/src/pages/feature/detail.tsx [hunk 4]
```diff
@@ -73,7 +74,8 @@ export const FeatureDetailPage: FC = memo(() => {
     }
   }, [featureId, dispatch, currentEnvironment]);
 
-  if (!feature || isLoading) {
+  if (!feature) {
+    // if (!feature || isLoading) {
```
Comment (cre8ivejp) [inline ui/web-v2/apps/admin/src/pages/feature/detail.tsx:78]: Please remove this line.

**Comment by cre8ivejp** [comment]
@bimalgrg519 

After adding all the flags, if you remove them, the `Add Prerequisite` button stays disabled. Even if you try to reload the page, or access another feature flag, the button stays disabled.

http://localhost:8000/development/features/test-tag-filter/targeting

File: ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx [hunk 5]
```diff
@@ -312,9 +312,7 @@ export const PrerequisiteInput: FC<PrerequisiteInputProps> = memo(
     };
 
     useEffect(() => {
-      if (prerequisites?.length > 0) {
-        dispatchListFeatures();
-      }
+      dispatchListFeatures();
```
Comment (cre8ivejp) [inline ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:315]: @bimalgrg519 

This way the dispatch will always trigger when the page is accessed.

We want to dispatch it when the page is loaded and it has prerequisites.
Or when it doesn't have prerequisites, we want to dispatch it when the `Add Prerequisites` button is clicked.

Comment (bimalgrg519) [reply ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:315]: @cre8ivejp  Yeah i agree with your points. But to make add Prerequisites button enable or disable at first, we need total numbers of feature flag. That's why we need to dispatch feature flag list api.

Comment (cre8ivejp) [reply ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:315]: @bimalgrg519 
Can't you can solve this by adding a condition inside of `disableAddPrerequisite`?

```ts
useEffect(() => {
  if (prerequisites.length > 0) {
    dispatchListFeatures();
  }
}, []);

const disableAddPrerequisite = useCallback(() => {
  if (prerequisites.length === 0) {
    return false;
  }
  return prerequisites.length === features.length - 1;
}, [prerequisites, features]);
```

Comment (bimalgrg519) [reply ui/web-v2/apps/admin/src/components/FeatureTargetingForm/index.tsx:315]: Thank you so much @cre8ivejp . Perfect solution.

**Review by cre8ivejp — DISMISSED** [review]
Thank you!

File: yarn.lock [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (cre8ivejp) [inline yarn.lock:1]: @bimalgrg519, we don't need this file.

Comment (cre8ivejp) [reply yarn.lock:1]: @bimalgrg519 can you take a look at the comments when you have time, please?

Comment (bimalgrg519) [reply yarn.lock]: @cre8ivejp I mistakenly added this file. I have removed it now.

**Review by cre8ivejp — APPROVED** [review]
👍

**Pull request merged**